File: frontend/src/types/gym.ts
export type GymAccess = 'public' | 'private' | 'secret';
export type SubscriptionType = 'free' | 'members_free' | 'paid' | 'tiered';
export type GymRole = 'owner' | 'co_owner' | 'trainer' | 'moderator' | 'member' | 'guest';

export interface OwnerData {
  user_id: string;
  username: string;
  display_name: string;
  avatar_url: string;
  role: string;
}

export interface GymCategory {
  id: string | number;
  name: string;
  display_name: string;
  icon: string;
  is_active: boolean;
}

export interface GymCategoryPricing {
  id?: string;
  category: string | number;
  category_name?: string;
  fee_per_day: number | null;
  fee_per_week: number | null;
  fee_per_month: number | null;
  fee_per_year: number | null;
  is_free: boolean;
}

export interface Gym {
  id: string;
  name: string;
  handle: string;
  description: string;
  logo_url: string;
  cover_url: string;
  category: string;
  categories: GymCategory[];
  access_type: GymAccess;
  subscription_type: SubscriptionType;
  monthly_fee_artifacts: Record<string, number> | null;
  join_fee_artifacts: Record<string, number> | null;
  category_pricing: GymCategoryPricing[];
  is_verified: boolean;
  is_reviews_enabled: boolean;
  is_donations_enabled: boolean;
  average_rating?: number;
  review_count?: number;
  recent_reviewers?: MemberData[];
  rules: string[];
  tags: string[];
  member_count: number;
  active_today: number;
  location_city: string;
  location_country: string;
  owner_data: OwnerData[];
  membership_role: GymRole | null;
  is_member: boolean;
  created_at: string;
  updated_at: string;
}

export interface MemberData {
  user_id: string;
  username: string;
  display_name: string;
  avatar_url: string;
  verification_status?: string;
}

export interface GymMembership {
  id: string;
  gym_id: string;
  member_id: string;
  role: GymRole;
  subscription_active: boolean;
  subscription_expires_at: string | null;
  member_data: MemberData;
  created_at: string;
}

export interface JoinRequest {
  id: string;
  gym_id: string;
  requester: string;
  requester_data: MemberData;
  message: string;
  status: 'pending' | 'approved' | 'rejected';
  reviewed_by: string | null;
  reviewed_at: string | null;
  created_at: string;
}

export interface GymInvite {
  id: string;
  gym_id: string;
  invited_user: string;
  invited_user_data: MemberData;
  invited_by: string;
  invited_by_data: { user_id: string; username: string; display_name: string };
  status: 'pending' | 'accepted' | 'declined';
  created_at: string;
}

export interface CityResult {
  place_id: string;
  city: string;
  country: string;
  description: string;
}

export interface GymSchedulePost {
  id: string;
  gym_id: string;
  author: string;
  author_data: MemberData;
  title: string;
  content: string;
  activity_type: string;
  custom_activity_type: string;
  location_mode: string;
  start_time: string | null;
  end_time: string | null;
  recurrence?: string;
  recurrence_end_date?: string | null;
  max_slots: number;
  enrollment_count?: number;
  is_enrolled?: boolean;
  created_at: string;
}

export interface GymReview {
  id: string;
  gym_id: string;
  reviewer: string;
  reviewer_data: MemberData;
  rating: number;
  comment: string;
  reply_text: string;
  replied_by: string | null;
  replied_by_data: MemberData | null;
  replied_at: string | null;
  created_at: string;
}

export interface GymDonation {
  id: string;
  gym_id: string;
  donor: string;
  donor_data: MemberData;
  amount: string | number;
  message: string;
  created_at: string;
}

File: buddy_up_flutter/lib/data/models/gym.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'gym.freezed.dart';
part 'gym.g.dart';

@freezed
abstract class OwnerData with _$OwnerData {
  const factory OwnerData({
    required String userId,
    required String username,
    required String displayName,
    required String avatarUrl,
    required String role,
  }) = _OwnerData;

  factory OwnerData.fromJson(Map<String, dynamic> json) => _$OwnerDataFromJson(json);
}

@freezed
abstract class MemberData with _$MemberData {
  const factory MemberData({
    required String userId,
    required String username,
    required String displayName,
    required String avatarUrl,
    @Default('none') String verificationStatus,
  }) = _MemberData;

  factory MemberData.fromJson(Map<String, dynamic> json) => _$MemberDataFromJson(json);
}

@freezed
abstract class GymCategory with _$GymCategory {
  const factory GymCategory({
    required String id,
    required String name,
    required String displayName,
    @Default('') String icon,
    @Default(true) bool isActive,
  }) = _GymCategory;

  factory GymCategory.fromJson(Map<String, dynamic> json) => _$GymCategoryFromJson(json);
}

@freezed
abstract class GymCategoryPricing with _$GymCategoryPricing {
  const factory GymCategoryPricing({
    String? id,
    required String category,
    String? categoryName,
    double? feePerDay,
    double? feePerWeek,
    double? feePerMonth,
    double? feePerYear,
    @Default(false) bool isFree,
  }) = _GymCategoryPricing;

  factory GymCategoryPricing.fromJson(Map<String, dynamic> json) =>
      _$GymCategoryPricingFromJson(json);
}

@freezed
abstract class Gym with _$Gym {
  const factory Gym({
    required String id,
    required String name,
    required String handle,
    @Default('') String description,
    @Default('') String logoUrl,
    @Default('') String coverUrl,
    @Default('') String category,
    @JsonKey(name: 'content_rating') @Default('general') String contentRating,
    @Default(<GymCategory>[]) List<GymCategory> categories,
    @Default('public') String accessType,
    @Default('free') String subscriptionType,
    @Default(false) bool isVerified,
    @Default(true) bool isReviewsEnabled,
    @Default(false) bool isDonationsEnabled,
    double? averageRating,
    @Default(0) int reviewCount,
    @Default(<MemberData>[]) List<MemberData> recentReviewers,
    @Default(<String>[]) List<String> rules,
    @Default(<String>[]) List<String> tags,
    @Default(0) int memberCount,
    @Default(0) int activeToday,
    @Default('') String locationCity,
    @Default('') String locationCountry,
    @Default(<OwnerData>[]) List<OwnerData> ownerData,
    String? membershipRole,
    @Default(false) bool isMember,
    required String createdAt,
    String? updatedAt,
  }) = _Gym;

  factory Gym.fromJson(Map<String, dynamic> json) => _$GymFromJson(json);
}

@freezed
abstract class GymMembership with _$GymMembership {
  const factory GymMembership({
    required String id,
    required String gymId,
    required String memberId,
    @Default('member') String role,
    @Default(false) bool subscriptionActive,
    String? subscriptionExpiresAt,
    required MemberData memberData,
    required String createdAt,
  }) = _GymMembership;

  factory GymMembership.fromJson(Map<String, dynamic> json) =>
      _$GymMembershipFromJson(json);
}

@freezed
abstract class JoinRequest with _$JoinRequest {
  const factory JoinRequest({
    required String id,
    required String gymId,
    required String requester,
    required MemberData requesterData,
    @Default('') String message,
    @Default('pending') String status,
    String? reviewedBy,
    String? reviewedAt,
    required String createdAt,
  }) = _JoinRequest;

  factory JoinRequest.fromJson(Map<String, dynamic> json) =>
      _$JoinRequestFromJson(json);
}

@freezed
abstract class GymInvite with _$GymInvite {
  const factory GymInvite({
    required String id,
    required String gymId,
    required String invitedUser,
    required MemberData invitedUserData,
    required String invitedBy,
    required Map<String, dynamic> invitedByData,
    @Default('pending') String status,
    required String createdAt,
  }) = _GymInvite;

  factory GymInvite.fromJson(Map<String, dynamic> json) => _$GymInviteFromJson(json);
}

@freezed
abstract class CityResult with _$CityResult {
  const factory CityResult({
    required String placeId,
    required String city,
    required String country,
    required String description,
  }) = _CityResult;

  factory CityResult.fromJson(Map<String, dynamic> json) => _$CityResultFromJson(json);
}

@freezed
abstract class GymSchedulePost with _$GymSchedulePost {
  const factory GymSchedulePost({
    required String id,
    required String gymId,
    required String author,
    required MemberData authorData,
    @Default('') String title,
    @Default('') String content,
    @Default('') String activityType,
    @Default('') String customActivityType,
    @Default('') String locationMode,
    String? startTime,
    String? endTime,
    String? recurrence,
    String? recurrenceEndDate,
    String? recurrenceDays,
    @Default(0) int maxSlots,
    @Default(0) int enrollmentCount,
    @Default(false) bool isEnrolled,
    required String createdAt,
  }) = _GymSchedulePost;

  factory GymSchedulePost.fromJson(Map<String, dynamic> json) =>
      _$GymSchedulePostFromJson(json);
}

@freezed
abstract class GymReview with _$GymReview {
  const factory GymReview({
    required String id,
    required String gymId,
    required String reviewer,
    required MemberData reviewerData,
    required int rating,
    @Default('') String comment,
    @Default('') String replyText,
    String? repliedBy,
    MemberData? repliedByData,
    String? repliedAt,
    required String createdAt,
  }) = _GymReview;

  factory GymReview.fromJson(Map<String, dynamic> json) => _$GymReviewFromJson(json);
}

@freezed
abstract class GymDonation with _$GymDonation {
  const factory GymDonation({
    required String id,
    required String gymId,
    required String donor,
    required MemberData donorData,
    required String amount,
    @Default('') String message,
    required String createdAt,
  }) = _GymDonation;

  factory GymDonation.fromJson(Map<String, dynamic> json) =>
      _$GymDonationFromJson(json);
}

@freezed
abstract class GymEvent with _$GymEvent {
  const factory GymEvent({
    required String id,
    required String gymId,
    required String title,
    @Default('') String description,
    String? startTime,
    String? endTime,
    @Default('') String location,
    required String createdAt,
  }) = _GymEvent;

  factory GymEvent.fromJson(Map<String, dynamic> json) => _$GymEventFromJson(json);
}

@freezed
abstract class CreateGymPayload with _$CreateGymPayload {
  const factory CreateGymPayload({
    required String name,
    required String handle,
    String? description,
    required String category,
    @JsonKey(name: 'content_rating') @Default('general') String contentRating,
    @Default(<String>[]) List<String> categoryIds,
    @Default('public') String accessType,
    @Default('free') String subscriptionType,
    String? locationCity,
    String? locationCountry,
    @Default(<String>[]) List<String> rules,
    @Default(<String>[]) List<String> tags,
    @Default(<GymCategoryPricing>[]) List<GymCategoryPricing> categoryPricing,
  }) = _CreateGymPayload;

  factory CreateGymPayload.fromJson(Map<String, dynamic> json) =>
      _$CreateGymPayloadFromJson(json);
}
